---
title: 裤子把人搞死？一个库引发的悲剧
comments: true
date: 2017-04-22 13:58:59
tags: [c, debug, ODBC, gdb]
categories: [技术分享]
---

> 穿个裤子，裤子破个洞，人应该没事吧？现实是人有可能挂掉！悲剧正在发生，一些破库经常搞挂服务器。

问题现象
-----------
服务器coredump，堆栈显示如下：


![**x**][1]
数据库宕记前日志显示如下：

![image_1be7j5vnq1iqh1ub1b0bcqm17oi9.png-3786.2kB][2]


通过数据库日志来看，这条语句为
```
select * from D5000.OP_ANA_STAT_SAMPLE_MONTH@kdb;
```
手动执行这条语句，并没有引发coredump。可见并不是一个频繁发生的问题。

堆栈信息可知，当时在执行一条pbe语句，在execute阶段通过dblink获取数据时，由于某种原因失败，在释放内存的时候core在了odbc驱动之中。栈顶函数为free，怀疑是odbc驱动内存释放不当导致。


问题分析
---------

### 错误行定位
由于odbc没有显示行号，需要根据反汇编来确定出错行的位置。根据堆栈信息，出错的指令地址位于0x2aba66c31370位置。

该处汇编显示为：

<center> ![image_1be76mofi1gucdfu1n8214dl17809.png-3387.8kB][3] </center>

当时的寄存器显示为：

<center> ![image_1be770h6n1jdqp6u12ja1ltp1cvvm.png-3299.7kB][4] </center>

通过反汇编定位到出错行为：511行。
```c
   491 void
   492 QR_free_memory(QResultClass *self)
   493 {
   494 ›   register int lf,
   495 ›   ›   ›   ›   row;
   496 ›   register TupleField *tuple = self->backend_tuples;
   497 ›   SQLLEN› ›   num_backend_rows = self->num_backend_rows;
   498 ›   int››   ›   num_fields = self->num_fields;
   499
   500 ›   mylog("QResult: free memory in %x, fcount = %d\n", self, num_backend_rows);
   501 ›   if (self->backend_tuples)
   502 ›   {
   503 ›   ›   for (row = 0; row < num_backend_rows; row++)
   504 ›   ›   {
   505 ›   ›   ›   mylog("row = %d, num_fields = %d\n", row, num_fields);
   506 ›   ›   ›   for (lf = 0; lf < num_fields; lf++)
   507 ›   ›   ›   {
   508 ›   ›   ›   ›   if (tuple[lf].value != NULL)
   509 ›   ›   ›   ›   {
   510 ›   ›   ›   ›   ›   mylog("free [lf=%d] %x\n", lf, tuple[lf].value);
   511 ›   ›   ›   ›   ›   free(tuple[lf].value);    ---出错行
   512 ›   ›   ›   ›   ›   tuple[lf].value = NULL;
   513 ›   ›   ›   ›   }
   514 ›   ›   ›   }
   515 ›   ›   ›   tuple += num_fields;›   /* next row */
   516 ›   ›   }
   517
   518 ›   ›   free(self->backend_tuples);
   519 ›   ›   //added by YJL 2003-03-18
   520 ›   ›   self->backend_tuples=NULL;
   521 ›   ›   self->count_backend_allocated = 0;
   522 ›   ›   self->backend_tuples = NULL;
   523 ›   }
```

free的地址为：0x286f1。

通过对比堆上地址来看:
堆地址范围：  0x2eb465c8
指令地址范围：0x5f0e9c 附近

显然free的地址0x286f1为无效地址。

### 错误行dblink.c:789
该处能够直接对应到行号，该行附近的代码如下：
```c
780
781 ›   ›   ›   tuplebuf = palloc(sizeof(char *) * query->colcount);
782
783 ›   ›   ›   while((rc = SQLFetch(query->hstmt)) != SQL_NO_DATA)
784 ›   ›   ›   {
785 ›   ›   ›   ›   HeapTuple›  tuple;
786 ›   ›   ›   ›   int         offset;
787 ›   ›   ›   ›
788 ›   ›   ›   ›   if (rc != SQL_SUCCESS && rc != SQL_SUCCESS_WITH_INFO)
789 ›   ›   ›   ›   ›   DBLinkReport(query, handle);
790
791 ›   ›   ›   ›   for(offset = 0; offset < query->fetchedrows; offset++)
792 ›   ›   ›   ›   {
793 ›   ›   ›   ›   ›   for(i = 0; i< query->colcount; i++)
794 ›   ›   ›   ›   ›   {
```
可见，SQLFetch函数失败，程序调用DBLinkReport报错退出。报错信息通过gdb能够看到如下内容：


![image_1bea3hfad80cat3mmr1u0q5ko2a.png-186.6kB][6]

报的错误为：`Error Fetching next group`

根据报错信息，对应到odbc代码中898行位置：
```c
in function
   886 ›   ›   ›   sprintf(fetch, "fetch %d in %s", fetch_size, self->cursor);
   887
   888 ›   ›   ›   mylog("next_tuple: sending actual fetch (%d) query '%s'\n", fetch_size, fetch);
   889
   890 ›   ›   ›   /* don't read ahead for the next tuple (self) ! */
   891 ›   ›   ›   qi.row_size = self->cache_size;
   892 ›   ›   ›   qi.result_in = self;
   893 ›   ›   ›   qi.cursor = NULL;
   894 ›   ›   ›   res = DBC_send_query(self->conn, fetch, &qi, CLEAR_RESULT_ON_ABORT);
   895 ›   ›   ›   if (res == NULL)
   896 ›   ›   ›   {
   897 ›   ›   ›   ›   self->status = KBRES_FATAL_ERROR;
   898 ›   ›   ›   ›   QR_set_message(self, "Error fetching next group.", 1);
   899 ›   ›   ›   ›   return FALSE;
   900 ›   ›   ›   }
   901 ›   ›   ›   self->inTuples = TRUE;
```
根据代码，可以看到当时`DBC_send_query`在执行`fetch %d in %s`的时候失败。

### 宕机前调用堆栈

目前看到的是宕记时候的堆栈，现在根据已知信息还原coredump前调用的堆栈：

<div class="mermaid">
graph TD
DBC_send_query   --失败返回NULL-->                                QR_next_tuple
QR_next_tuple    --报错信息Error fetching_next group 返回FALSE--> STMT_Fetch
STMT_Fetch       --SQLSTATE设为HY000返回SQLERROR-->               MY_ExtendedFetch
MY_ExtendedFetch --返回SQLERROR-->                                SQLFetch
SQLFetch         --返回SQLERROR-->                                dblink_record
dblink_record    --调用-->                                        DBLinkReport
DBLinkReport     --调用-->                                        SQLFreeHandle
SQLFreeHandle    --经过若干调用-->                                QR_free_memory
</div>


问题重现
-----------
仿照上一节的推理，如果DBC_send_query失败，问题会不会重现？
那么在DBC_send_query处打断点，接下来c，直到有fetch 1024的时候，强制DBC_send_query返回0
```
(gdb) b DBC_send_query
Function "DBC_send_query" not defined.
Make breakpoint pending on future shared library load? (y or [n]) y
Breakpoint 1 (DBC_send_query) pending.
(gdb) c
Continuing.
[New Thread 0x40a74940 (LWP 15989)]
[Switching to Thread 0x42332940 (LWP 15981)]

Breakpoint 1, DBC_send_query (self=0x9cf5e40, query=0x2aaade6554d9 " ", qi=0x0, flag=1) at connection.c:3820
3820		QResultClass *cmdres = NULL,
(gdb) c
Continuing.

Breakpoint 1, DBC_send_query (self=0x9cf5e40, query=0x9ebed80 "set DateStyle to 'ISO'", qi=0x0, flag=0) at connection.c:3820
3820		QResultClass *cmdres = NULL,
(gdb)
Continuing.


Breakpoint 1, DBC_send_query (self=0x9cf5e40, query=0x9ee1240 "declare SQL_CUR0x9ec1c20 scroll cursor for SELECT * FROM \"TB1\" WHERE 1 = 0", qi=0x0, flag=4) at connection.c:3820
3820		QResultClass *cmdres = NULL,
(gdb)
Continuing.

Breakpoint 1, DBC_send_query (self=0x9cf5e40, query=0x2aaade6548e5 "BEGIN", qi=0x0, flag=1) at connection.c:3820
3820		QResultClass *cmdres = NULL,
(gdb)
Continuing.

Breakpoint 1, DBC_send_query (self=0x9cf5e40, query=0x4229ee50 "fetch 0 in SQL_CUR0x9ec1c20", qi=0x4229eee0, flag=0) at connection.c:3820
3820		QResultClass *cmdres = NULL,
(gdb)
Continuing.

Breakpoint 1, DBC_send_query (self=0x9cf5e40, query=0x4229f160 "close SQL_CUR0x9ec1c20", qi=0x0, flag=1) at connection.c:3820
3820		QResultClass *cmdres = NULL,
(gdb)
Continuing.

Breakpoint 1, DBC_send_query (self=0x9cf5e40, query=0x9ec1ba0 "declare SQL_CUR0x9ee1240 scroll cursor for SELECT * FROM \"TB1\"", qi=0x0, flag=0) at connection.c:3820
3820		QResultClass *cmdres = NULL,
(gdb)
Continuing.

Breakpoint 1, DBC_send_query (self=0x9cf5e40, query=0x4229ebc0 "fetch 0 in SQL_CUR0x9ee1240", qi=0x4229ec50, flag=0) at connection.c:3820
3820		QResultClass *cmdres = NULL,
(gdb)
Continuing.

Breakpoint 1, DBC_send_query (self=0x9cf5e40, query=0x4229cad0 "fetch 1024 in SQL_CUR0x9ee1240", qi=0x4229cab0, flag=1) at connection.c:3820
3820		QResultClass *cmdres = NULL,
(gdb) return 0

```
接下来就会core掉
```
#0  0x0000003a93e7273e in free () from /lib64/libc.so.6
#1  0x00002aaade575312 in QR_free_memory (self=0x1b7a17a0) at qresult.c:511
#2  0x00002aaade574c04 in QR_Destructor (self=0x1b7a17a0, closeCursor=1) at qresult.c:296
#3  0x00002aaade58d5f2 in STMT_Destructor (self=0x1b7bfe30) at statement.c:507
#4  0x00002aaade58cec4 in MY_FreeStmt (hstmt=0x1b7bfe30, fOption=1) at statement.c:313
#5  0x00002aaade59a1ad in SQLFreeHandle (HandleType=3, Handle=0x1b7bfe30) at odbcapi30.c:252
#6  0x0000003a94e170ed in __SQLFreeHandle (handle_type=<value optimized out>, handle=0x1b7c1650) at SQLFreeHandle.c:435
#7  0x0000003a94e1768c in SQLFreeHandle (handle_type=17, handle=0x0) at SQLFreeHandle.c:598
#8  0x0000000000b3591b in FreeQueryHandle (conn=0x1b4117a8) at dblinkutl.c:161
#9  0x0000000000b36c1c in DBLinkReport (query=0x1b79e8c0, cell=0x1b412348) at dblinkerr.c:68
#10 0x0000000000b34f5a in dblink_record (fcinfo=0x417853c0) at dblink.c:789
#11 0x00000000006b3b44 in ExecMakeTableFunctionResult (funcexpr=0x1b79e240, econtext=0x1b79de10, expectedDesc=0x1b79e0d0, returnDesc=0x41785820) at execQual.c:1431
#12 0x00000000006d055f in FunctionNext (node=0x1b79dcf0) at nodeFunctionscan.c:68
#13 0x00000000006bd964 in ExecScan (node=0x1b79dcf0, accessMtd=0x6d04f8 <FunctionNext>) at execScan.c:70
```

那么可以说明我们目前的判断基本正确。出现故障的原因更加清晰了：当时在执行一条pbe语句，在execute阶段通过dblink获取数据时，由于DBC_send_query失败，释放内存不当导致。接下来我们需要找到那块不当释放的内存。



问题根因
------------
我们找到backend_tuples分配内存的地方：
```
self->backend_tuples = (TupleField *) malloc(self->num_fields * sizeof(TupleField) * row);
```
TupleField定义如下：
```
/*› Used by backend data AND manual result sets */
struct TupleField_
{
›   Int4›   ›   len;›   ›   ›   /* length of the current Tuple */
›   void›      *value;› ›   ›   /* an array representing the value */
};
```

可见是分配了列数×行数这么大小的空间，可以看作一个表格：
```
backend_tuples
|---------------|---------------|---------------|---------------|
|tuple[0].len   |tuple[1].len   |tuple[2].len   |tuple[3].len   |
|tuple[0].value |tuple[1].value |tuple[2].value |tuple[3].value | row1
|---------------|---------------|---------------|---------------|
|tuple[0].len   |tuple[1].len   |tuple[2].len   |tuple[3].len   |
|tuple[0].value |tuple[1].value |tuple[2].value |tuple[3].value | row2
|---------------|---------------|---------------|---------------|
.
.
.
```
每个格子都是一个TupleField
结合`QR_free_memory`这个函数来看
```c
   491 void
   492 QR_free_memory(QResultClass *self)
   493 {
   494 ›   register int lf,
   495 ›   ›   ›   ›   row;
   496 ›   register TupleField *tuple = self->backend_tuples;
   497 ›   SQLLEN› ›   num_backend_rows = self->num_backend_rows;
   498 ›   int››   ›   num_fields = self->num_fields;
   499
   500 ›   mylog("QResult: free memory in %x, fcount = %d\n", self, num_backend_rows);
   501 ›   if (self->backend_tuples)
   502 ›   {
   503 ›   ›   for (row = 0; row < num_backend_rows; row++)
   504 ›   ›   {
   505 ›   ›   ›   mylog("row = %d, num_fields = %d\n", row, num_fields);
   506 ›   ›   ›   for (lf = 0; lf < num_fields; lf++)
   507 ›   ›   ›   {
   508 ›   ›   ›   ›   if (tuple[lf].value != NULL)
   509 ›   ›   ›   ›   {
   510 ›   ›   ›   ›   ›   mylog("free [lf=%d] %x\n", lf, tuple[lf].value);
   511 ›   ›   ›   ›   ›   free(tuple[lf].value);    ---出错行
   512 ›   ›   ›   ›   ›   tuple[lf].value = NULL;
   513 ›   ›   ›   ›   }
   514 ›   ›   ›   }
   515 ›   ›   ›   tuple += num_fields;›   /* next row */

```
是需要挨个释放 tuple->value这个指针指向的地址。释放多少个tuple呢？是根据num_backend_rows来判断的。
那么就存在如下2个问题：

1. malloc分配的空间，必须置空0，否则tuple.value是一个垃圾值。
2. num_backend_rows 是否准确？过小则造成内存泄漏，过大则会引起tuple.value指向未知内存。
3. tuple->value是否指向有效地址？比如指向已经释放过的地址（释放了必须置NULL），或者无效地址（比如栈空间）。

通过看代码，确实存在问题，比如realloc了内存忘记清空，num_backend_rows在读取成功之前就存在递增的情况，如果DBC_send_query失败那么就会有上面第二个问题出现。
另外，tuple->value确实存在指向栈上空间的情况。


查找value指针指向的内存：

1、绝大部分地方value在free之后都置NULL了，但是在TL_Destructor以及STMT_pos_reload_needed函数中，value释放过后没置NULL，可能引起多次释放的问题。

2、在QR_read_tuple函数中，buffer指向的位置是由如下逻辑确定的：

```c

            char    tidoidbuf[32];
            if (field_lf >= effective_cols)
                  buffer = tidoidbuf;   --buffer指向stack位置
            else
             {
                buffer = (char *) malloc(len + 1); --buffer指向heap位置
                if(buffer == NULL)
                    return FALSE;
            }
```

如果满足field_lf >= effective_cols那么是指向stack上的内存，否则是指向heap上的内存。

```c
            if (field_lf >= effective_cols && self->haskeyset)
            {
                       }
            else
            {
                this_tuplefield[field_lf].len = len;
                this_tuplefield[field_lf].value = buffer;   --buffer有可能指向stack
                      }
```

但是如上代码在实际使用过程中，当`self->haskeyset = false`的时候，就会返回stack上的内存。




问题修复
------------
主要是对上面提出的3个问题进行修复：
1. 记录malloc,realloc分配内存的范围，检查tuple是否在这个范围之中。
2. malloc,realloc之后，用memset置空内存，防止垃圾值影响tuple.value。
3. 确保tuple.value不会指向栈上的内存。


修复过后，在DBC_send_query失败之后就不会core了，错误正常返回给客户端。
```
TEST=# SELECT count(*) from tb1@kdb;
ERROR:  error in database link "KDB", SQLSTATE=[HY000] Error fetching next group.
STATEMENT:  SELECT count(*) from tb1@kdb;
LOG:  statement: ROLLBACK
ERROR:  error in database link "KDB", SQLSTATE=[HY000] Error fetching next group.
TEST=#
```


  [1]: http://static.zybuluo.com/shenyuflying/kxnvjyy0kkifws502en6rkvu/image_1bea3g6li19431k0dpq61fag16pn1t.png
  [2]: http://static.zybuluo.com/shenyuflying/7ut8qfw12wql4rrb5x1hxl54/image_1be7j5vnq1iqh1ub1b0bcqm17oi9.png
  [3]: http://static.zybuluo.com/shenyuflying/0ys6sy1fy85kiq8eofdgpu4m/image_1be76mofi1gucdfu1n8214dl17809.png
  [4]: http://static.zybuluo.com/shenyuflying/gdrztnkejhms20aij08fnaxd/image_1be770h6n1jdqp6u12ja1ltp1cvvm.png
  [5]: http://static.zybuluo.com/shenyuflying/w5awjqpcq1a4x5bgnlde0tsj/image_1be77qi2g1cbe1bd7u4i1qk1s8713.png
  [6]: http://static.zybuluo.com/shenyuflying/hprhegmbg94f8vat7jggh3jq/image_1bea3hfad80cat3mmr1u0q5ko2a.png


如果还想看到更多此类文章，请移步到[小宇的博客](http://shenyu.wiki)。
